[Docs] sylius performance boost (#12873)
Sylius/Sylius: docs/performance/database-indexes.rst

Database indexes
================

Indexing tables allows you to decrease fetching time from database.

As an example lets take a look on customers list

> Default index page is sorted by registration date, to create table index all you need to do is modify `Customer` Entity and add index using annotations.

Reviewer: you should add here snippet with example how this annotation should look.

Database indexes
================

Indexing tables allows you to decrease fetching time from database.

As an example lets take a look on customers list

Default index page is sorted by registration date, to create table index on this page all you need to do is one database query:

.. code-block:: mysql

>     CREATE INDEX sylius_customer_created_at_index ON sylius_customer (created_at DESC);

Reviewer 2: IMHO such indexes should be set already in Doctrine entity declaration.

Database indexes
================

Indexing tables allows you to decrease fetching time from database.

As an example lets take a look on customers list

> Default index page is sorted by registration date, to create table index all you need to do is modify `src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Customer.orm.xml` and add

Reviewer 3: The problem with this line is that developers reading this doc will not be able to edit the `src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Customer.orm.xml` file - as it will be part of vendor code. This change should be done with annotations here: https://github.com/Sylius/Sylius-Standard/blob/1.10/src/Entity/Customer/Customer.php#L12

Can we also link to doctrine docs? https://www.doctrine-project.org/projects/doctrine-orm/en/latest/reference/annotations-reference.html#annref_haslifecyclecallbacks

.. code-block:: xml

    ...
    <indexes>
        <index name="created_at_index" columns="created_at" />
    </indexes>

then run `bin/console doctrine:migrations:diff` to create migration with proper index.

> 

Reviewer 3: ```suggestion

This should be considered for the most common sorting in your application.

```

Database indexes
================

> Indexing tables allows you to decrease fetching time from database.

Reviewer 3: ```suggestion
Indexing tables allow you to decrease fetching time from the database.
```

> As an example lets take a look on customers list

Reviewer 3: ```suggestion
As an example, let's take a look at the customers' list.
```

> Default index page is sorted by registration date, to create table index all you need to do is modify `Customer` Entity and add index using annotations.

Reviewer 3: ```suggestion
The default index page is sorted by registration date, to create a table index all you need to do is modify `\App\Entity\Customer\Customer` entity and add the index using annotations. This class can be found at `src/Entity/Customer/Customer.php`. Once open, paste the following code:
```

Reviewer 4: @lchrusciel why once the slashes are forward and once backward?

Reviewer 3: one was the file path, while the other FQCN

`ORM annotations <https://www.doctrine-project.org/projects/doctrine-orm/en/latest/reference/annotations-reference.html#annref_haslifecyclecallbacks>`_

.. code-block:: php

>     /**
>      * @Entity
>      * @Table(name="customer",indexes={@Index(name="created_at_index", fields={"created_at"})})
>      */

Reviewer 3: ```suggestion
     <?php

    // src/Entity/Customer/Customer.php
    
    declare(strict_types=1);
    
    namespace App\Entity\Customer;
    
    use Doctrine\ORM\Mapping as ORM;
    use Sylius\Component\Core\Model\Customer as BaseCustomer;
    
    /**
     * @ORM\Entity
     * @ORM\Table(name="customer",indexes={@Index(name="created_at_index", fields={"created_at"})})
     */
    class Customer extends BaseCustomer
    {
    }

```

Reviewer 4: `// src/Entity/Customer/Customer.php` is not even needed as we have namespace.

Database indexes
================

Indexing tables allow you to decrease fetching time from the database.

As an example, let's take a look at the customers' list.

> The default index page is sorted by registration date, to create a table index all you need to do is modify `\App\Entity\Customer\Customer` entity and add the index using annotations. This class can be found at `src/Entity/Customer/Customer.php`. Once open, paste the following code:

Reviewer 5: "This class can be found at `src/Entity/Customer/Customer.php`" is it needed? I think we can assume developers would know how to find a class by its FQCN :)

Reviewer 5: `Once open, paste the following code` - it is not exactly true. One can already have some custom code in the class, so we would not like to override it. I would write sth like

```rst
In this file, add indexes attribute into the table configuration:
`@ORM\Table(name="customer", indexes={@Index(name="created_at_index", fields={"created_at"})})`.
Your class should now look like this:
```

.. code-block:: php

    <?php

    declare(strict_types=1);

    namespace App\Entity\Customer;

    use Doctrine\ORM\Mapping as ORM;
    use Sylius\Component\Core\Model\Customer as BaseCustomer;

    /**
     * @ORM\Entity
     * @ORM\Table(name="customer",indexes={@Index(name="created_at_index", fields={"created_at"})})
     */
    class Customer extends BaseCustomer
    {
    }

.. note::

    You can learn more here about `ORM annotations <https://www.doctrine-project.org/projects/doctrine-orm/en/latest/reference/annotations-reference.html#annref_haslifecyclecallbacks>`_

This should be considered for the most common sorting in your application.

> Using this solution you can increase speed of customer listing by around 10%.

Reviewer 5: I'm missing the explanation why it's not done out-of-the-box. When I read it I feel like "hey, folks, if it's so awesome why it's not done already in Sylius itself?" 🐎

Reviewer 5: Something like [this](https://github.com/Sylius/Sylius/pull/12873/files#diff-35c9ef260c46b3571ef272962775f92961a61e7307822f522fe4f4d3406c8eb9R22)

Author: changed it to a note with more info

Database indexes
================

Indexing tables allow you to decrease fetching time from the database.

As an example, let's take a look at the customers' list.

> The default index page is sorted by registration date, to create a table index all you need to do is modify `\App\Entity\Customer\Customer` entity and add the index using annotations.

Reviewer 6: ```suggestion
The default index page is sorted by registration date, to create a table index all you need to do is modify `App\Entity\Customer\Customer` entity and add the index using annotations.
```

As an example, let's take a look at the customers' list.

The default index page is sorted by registration date, to create a table index all you need to do is modify `\App\Entity\Customer\Customer` entity and add the index using annotations.

In this file, add indexes attribute into the table configuration:

.. code-block:: php

>     `@ORM\Table(name="customer", indexes={@Index(name="created_at_index", fields={"created_at"})})`.

Reviewer 2: ```suggestion
    `@ORM\Table(name="customer", indexes={@ORM\Index(name="created_at_index", fields={"created_at"})})`.
```
